### ISSUE TITLE
NullPointerException for `bazel test` when using `--spawn_strategy=remote`

### ISSUE THREAD

**Issue by razvanm** [bazelbuild/bazel#3913]
### Description of the problem / feature request / question:

My `~/.bazelrc` is the following:

```
$ cat ~/.bazelrc
startup --host_jvm_args=-Dbazel.DigestFunction=SHA1
build --spawn_strategy=remote
build --remote_rest_cache=http://127.0.0.1/cache
```

The error when running `bazel test` on the problematic target:

```
204	java.lang.RuntimeException: Unrecoverable error while evaluating node 'ACTION_EXECUTION:ActionLookupData{actionLookupNode=CONFIGURED_TARGET://<<edited>> 5b43e65b4db3f167ea10aef27250fef5 (18566788 1346308532), actionIndex=11}' (requested by nodes '<<edited>>/test.cache_status //<<edited>> 5b43e65b4db3f167ea10aef27250fef5 (2114509961 1346308532)')
205		at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:413)
206		at com.google.devtools.build.lib.concurrent.AbstractQueueVisitor$WrappedRunnable.run(AbstractQueueVisitor.java:352)
207		at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
208		at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
209		at java.lang.Thread.run(Thread.java:748)
210	Caused by: java.lang.NullPointerException
211		at com.google.common.hash.HashCode.fromBytes(HashCode.java:247)
212		at com.google.devtools.build.lib.remote.Digests.buildDigest(Digests.java:98)
213		at com.google.devtools.build.lib.remote.Digests.getDigestFromInputCache(Digests.java:108)
214		at com.google.devtools.build.lib.remote.TreeNodeRepository.getOrComputeDirectory(TreeNodeRepository.java:317)
215		at com.google.devtools.build.lib.remote.TreeNodeRepository.computeMerkleDigests(TreeNodeRepository.java:347)
216		at com.google.devtools.build.lib.remote.TreeNodeRepository.computeMerkleDigests(TreeNodeRepository.java:345)
217		at com.google.devtools.build.lib.remote.TreeNodeRepository.computeMerkleDigests(TreeNodeRepository.java:345)
218		at com.google.devtools.build.lib.remote.TreeNodeRepository.computeMerkleDigests(TreeNodeRepository.java:345)
219		at com.google.devtools.build.lib.remote.TreeNodeRepository.computeMerkleDigests(TreeNodeRepository.java:345)
220		at com.google.devtools.build.lib.remote.TreeNodeRepository.computeMerkleDigests(TreeNodeRepository.java:345)
221		at com.google.devtools.build.lib.remote.TreeNodeRepository.computeMerkleDigests(TreeNodeRepository.java:345)
222		at com.google.devtools.build.lib.remote.TreeNodeRepository.computeMerkleDigests(TreeNodeRepository.java:345)
223		at com.google.devtools.build.lib.remote.TreeNodeRepository.computeMerkleDigests(TreeNodeRepository.java:345)
224		at com.google.devtools.build.lib.remote.TreeNodeRepository.computeMerkleDigests(TreeNodeRepository.java:345)
225		at com.google.devtools.build.lib.remote.TreeNodeRepository.computeMerkleDigests(TreeNodeRepository.java:345)
226		at com.google.devtools.build.lib.remote.TreeNodeRepository.computeMerkleDigests(TreeNodeRepository.java:345)
227		at com.google.devtools.build.lib.remote.RemoteSpawnCache.lookup(RemoteSpawnCache.java:82)
228		at com.google.devtools.build.lib.exec.AbstractSpawnStrategy.exec(AbstractSpawnStrategy.java:90)
229		at com.google.devtools.build.lib.exec.AbstractSpawnStrategy.exec(AbstractSpawnStrategy.java:63)
230		at com.google.devtools.build.lib.exec.StandaloneTestStrategy.executeTest(StandaloneTestStrategy.java:322)
231		at com.google.devtools.build.lib.exec.StandaloneTestStrategy.executeTestAttempt(StandaloneTestStrategy.java:276)
232		at com.google.devtools.build.lib.exec.StandaloneTestStrategy.exec(StandaloneTestStrategy.java:143)
233		at com.google.devtools.build.lib.rules.test.ExclusiveTestStrategy.exec(ExclusiveTestStrategy.java:44)
234		at com.google.devtools.build.lib.analysis.test.TestRunnerAction.execute(TestRunnerAction.java:663)
235		at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor.executeActionTask(SkyframeActionExecutor.java:850)
236		at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor.prepareScheduleExecuteAndCompleteAction(SkyframeActionExecutor.java:793)
237		at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor.access$900(SkyframeActionExecutor.java:107)
238		at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$ActionRunner.call(SkyframeActionExecutor.java:661)
239		at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor$ActionRunner.call(SkyframeActionExecutor.java:618)
240		at java.util.concurrent.FutureTask.run(FutureTask.java:266)
241		at com.google.devtools.build.lib.skyframe.SkyframeActionExecutor.executeAction(SkyframeActionExecutor.java:404)
242		at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.checkCacheAndExecuteIfNeeded(ActionExecutionFunction.java:440)
243		at com.google.devtools.build.lib.skyframe.ActionExecutionFunction.compute(ActionExecutionFunction.java:201)
244		at com.google.devtools.build.skyframe.AbstractParallelEvaluator$Evaluate.run(AbstractParallelEvaluator.java:338)
245		... 4 more
246	
```

### If possible, provide a minimal example to reproduce the problem:

I don't have a small repro yet.

### Environment info
* Operating System:

```
$ uname -a
Linux bluephase 4.9.0-4-amd64 #1 SMP Debian 4.9.51-1 (2017-09-28) x86_64 GNU/Linux
$
```

* Bazel version (output of `bazel info release`):

```
$ bazel info release
.......
release 0.6.1
```

**Comment by ola-rozenfeld** [comment]
I still don't know the root cause, but I have a patch in review that will at least output a better error message (what input file was missing from local cache).

**Comment by dslomov** [comment]
Not sure what the right category is for remote execution. @damienmg wdyt?

**Comment by buchgr** [comment]
@dslomov service APIs is correct.

**Comment by buchgr** [comment]
`Metadata.getDigest()` returns `null`. Grepping through the code base suggests that this seems to be the case for middleman artifacts and directories.

**Comment by ola-rozenfeld** [comment]
A workaround - the reason for the bug is rules that have an entire directory as input, for example data=["somedir"]. If you change it to data=glob(["somedir/**"]), this will work.

I haven't realized until now that Bazel even supports this -- I will need to recursively expand directory inputs in the remote execution code for this. Alternatively, we could just say this is WAI and data should always refer to a collection of files, e.g. using a glob, rather than a directory. I will find out and update.

**Comment by ola-rozenfeld** [comment]
From @ulfjack :

I've talked with Lukacs about this a while ago and the basic plan we came up with was to support (and enforce) this usage in a specific way. Unfortunately, I never got around to writing up the plan.

There are three parts to the plan:
- allow a trailing '/' character in labels; mark those as special artifacts (source tree artifact?)
- enforce consistency between trailing '/' characters being there or not and the underlying file being a directory or not (this allows us to reject directories in the analysis phase in places where we can't support them; we don't want to stat the file in the analysis phase since doing so can be very costly)
- automatically run a 'glob' on directory artifacts before action execution for any action inputs that are directories (this'll register the dependencies in Skyframe, which fixes the current correctness issue); this may or may not use the same code path as glob(), but it's executed in the execution phase, not the analysis phase

The reason for not automatically converting it to glob during loading is that we don't want to pay the price for globbing, but it comes at the expense of having yet another special case for artifacts.

------------------------------------

This has been a plan for years and is not going to get done any time soon. For now, I will just throw a clear error in this case; as mentioned, directory data dependencies should be avoided in favor of explicit glob calls.
